File: blepro/src/main/java/com/lepu/blepro/ext/ventilator/WarningSetting.java
package com.lepu.blepro.ext.ventilator;

public class WarningSetting {

    private int type;
    private WarningApnea warningApnea;
    private WarningLeak warningLeak;
    private WarningVt warningVt;
    private WarningVentilation warningVentilation;
    private WarningRrHigh warningRrHigh;
    private WarningRrLow warningRrLow;
    private WarningSpo2Low warningSpo2Low;
    private WarningHrHigh warningHrHigh;
    private WarningHrLow warningHrLow;

    public int getType() {
        return type;
    }

    public void setType(int type) {
        this.type = type;
    }

    public WarningApnea getWarningApnea() {
        return warningApnea;
    }

    public void setWarningApnea(WarningApnea warningApnea) {
        this.warningApnea = warningApnea;
    }

    public WarningLeak getWarningLeak() {
        return warningLeak;
    }

    public void setWarningLeak(WarningLeak warningLeak) {
        this.warningLeak = warningLeak;
    }

    public WarningVt getWarningVt() {
        return warningVt;
    }

    public void setWarningVt(WarningVt warningVt) {
        this.warningVt = warningVt;
    }

    public WarningVentilation getWarningVentilation() {
        return warningVentilation;
    }

    public void setWarningVentilation(WarningVentilation warningVentilation) {
        this.warningVentilation = warningVentilation;
    }

    public WarningRrHigh getWarningRrHigh() {
        return warningRrHigh;
    }

    public void setWarningRrHigh(WarningRrHigh warningRrHigh) {
        this.warningRrHigh = warningRrHigh;
    }

    public WarningRrLow getWarningRrLow() {
        return warningRrLow;
    }

    public void setWarningRrLow(WarningRrLow warningRrLow) {
        this.warningRrLow = warningRrLow;
    }

    public WarningSpo2Low getWarningSpo2Low() {
        return warningSpo2Low;
    }

    public void setWarningSpo2Low(WarningSpo2Low warningSpo2Low) {
        this.warningSpo2Low = warningSpo2Low;
    }

    public WarningHrHigh getWarningHrHigh() {
        return warningHrHigh;
    }

    public void setWarningHrHigh(WarningHrHigh warningHrHigh) {
        this.warningHrHigh = warningHrHigh;
    }

    public WarningHrLow getWarningHrLow() {
        return warningHrLow;
    }

    public void setWarningHrLow(WarningHrLow warningHrLow) {
        this.warningHrLow = warningHrLow;
    }

    @Override
    public String toString() {
        return "WarningSetting{" +
                "type=" + type +
                ", warningApnea=" + warningApnea +
                ", warningLeak=" + warningLeak +
                ", warningVt=" + warningVt +
                ", warningVentilation=" + warningVentilation +
                ", warningRrHigh=" + warningRrHigh +
                ", warningRrLow=" + warningRrLow +
                ", warningSpo2Low=" + warningSpo2Low +
                ", warningHrHigh=" + warningHrHigh +
                ", warningHrLow=" + warningHrLow +
                '}';
    }

    // 报警提示：呼吸暂停
    public class WarningApnea {
        private int apnea;  // 0: Off 范围：10s/20s/30s；级别：高

        public int getApnea() {
            return apnea;
        }

        public void setApnea(int apnea) {
            this.apnea = apnea;
        }

        @Override
        public String toString() {
            return "WarningApnea{" +
                    "apnea=" + apnea +
                    '}';
        }
    }
    // 报警提示：漏气量高
    public class WarningLeak {
        private int high;  // 0: Off 范围：15s/30s/45s/60s； 级别：中

        public int getHigh() {
            return high;
        }

        public void setHigh(int high) {
            this.high = high;
        }

        @Override
        public String toString() {
            return "WarningLeak{" +
                    "high=" + high +
                    '}';
        }
    }
    // 报警提示：潮气量低
    public class WarningVt {
        private int low;  // 0: Off 范围：200-2000ml；步进：10ml；级别：中

        public int getLow() {
            return low;
        }

        public void setLow(int low) {
            this.low = low;
        }

        @Override
        public String toString() {
            return "WarningVt{" +
                    "low=" + low +
                    '}';
        }
    }
    // 报警提示：分钟通气量低
    public class WarningVentilation {
        private int low;  // 0: off 范围：1-25L/min；步进：1L/min；级别：中

        public int getLow() {
            return low;
        }

        public void setLow(int low) {
            this.low = low;
        }

        @Override
        public String toString() {
            return "WarningVentilation{" +
                    "low=" + low +
                    '}';
        }
    }
    // 报警提示：呼吸频率高
    public class WarningRrHigh {
        private int high;  // 0:Off 范围：1-60bpm；步进：1bpm；级别：中

        public int getHigh() {
            return high;
        }

        public void setHigh(int high) {
            this.high = high;
        }

        @Override
        public String toString() {
            return "WarningRrHigh{" +
                    "high=" + high +
                    '}';
        }
    }
    // 报警提示：呼吸频率低
    public class WarningRrLow {
        private int low;  // 0:Off 范围：1-60bpm；步进：1bpm；级别：中

        public int getLow() {
            return low;
        }

        public void setLow(int low) {
            this.low = low;
        }

        @Override
        public String toString() {
            return "WarningRrLow{" +
                    "low=" + low +
                    '}';
        }
    }
    // 报警提示：血氧饱和度低
    public class WarningSpo2Low {
        private int low;  // 0:Off 范围：80-95%；步进：1%；级别：中

        public int getLow() {
            return low;
        }

        public void setLow(int low) {
            this.low = low;
        }

        @Override
        public String toString() {
            return "WarningSpo2Low{" +
                    "low=" + low +
                    '}';
        }
    }
    // 报警提示：脉率/心率高
    public class WarningHrHigh {
        private int high;  // 0:Off 范围：100-240bpm；步进：10bpm；级别：中

        public int getHigh() {
            return high;
        }

        public void setHigh(int high) {
            this.high = high;
        }

        @Override
        public String toString() {
            return "WarningHrHigh{" +
                    "high=" + high +
                    '}';
        }
    }
    // 报警提示：脉率/心率低
    public class WarningHrLow {
        private int low;  // 0:Off 范围：30-70bpm；步进：5bpm；级别：中

        public int getLow() {
            return low;
        }

        public void setLow(int low) {
            this.low = low;
        }

        @Override
        public String toString() {
            return "WarningHrLow{" +
                    "low=" + low +
                    '}';
        }
    }
}

File: blepro/src/main/java/com/lepu/blepro/ble/data/r20/WarningSetting.kt
package com.lepu.blepro.ble.data.r20

import com.lepu.blepro.utils.ByteUtils.byte2UInt
import com.lepu.blepro.utils.int2ByteArray
import com.lepu.blepro.utils.toUInt

class WarningSetting() {

    var type = 0
    // reserved 3
    var warningApnea = WarningApnea()
    var warningLeak = WarningLeak()
    var warningVt = WarningVt()
    var warningVentilation = WarningVentilation()
    var warningRrHigh = WarningRrHigh()
    var warningRrLow = WarningRrLow()
    var warningSpo2Low = WarningSpo2Low()
    var warningHrHigh = WarningHrHigh()
    var warningHrLow = WarningHrLow()

    constructor(bytes: ByteArray) : this() {
        var index = 0
        warningApnea = WarningApnea(bytes.copyOfRange(index, index+1))
        index++
        warningLeak = WarningLeak(bytes.copyOfRange(index, index+1))
        index++
        warningVt = WarningVt(bytes.copyOfRange(index, index+1))
        index++
        warningVentilation = WarningVentilation(bytes.copyOfRange(index, index+1))
        index++
        warningRrHigh = WarningRrHigh(bytes.copyOfRange(index, index+1))
        index++
        warningRrLow = WarningRrLow(bytes.copyOfRange(index, index+1))
        index++
        warningSpo2Low = WarningSpo2Low(bytes.copyOfRange(index, index+1))
        index++
        warningHrHigh = WarningHrHigh(bytes.copyOfRange(index, index+1))
        index++
        warningHrLow = WarningHrLow(bytes.copyOfRange(index, index+1))
    }

    fun getDataBytes(): ByteArray {
        var data = byteArrayOf(type.toByte())
            .plus(ByteArray(3))
        when (type) {
            0 -> {
                data = data.plus(warningApnea.getDataBytes())
                    .plus(warningLeak.getDataBytes())
                    .plus(warningVt.getDataBytes())
                    .plus(warningVentilation.getDataBytes())
                    .plus(warningRrHigh.getDataBytes())
                    .plus(warningRrLow.getDataBytes())
                    .plus(warningSpo2Low.getDataBytes())
                    .plus(warningHrHigh.getDataBytes())
                    .plus(warningHrLow.getDataBytes())
            }
            1 -> {
                data = data.plus(warningApnea.getDataBytes())
            }
            2 -> {
                data = data.plus(warningLeak.getDataBytes())
            }
            3 -> {
                data = data.plus(warningVt.getDataBytes())
            }
            4 -> {
                data = data.plus(warningVentilation.getDataBytes())
            }
            5 -> {
                data = data.plus(warningRrHigh.getDataBytes())
            }
            6 -> {
                data = data.plus(warningRrLow.getDataBytes())
            }
            7 -> {
                data = data.plus(warningSpo2Low.getDataBytes())
            }
            8 -> {
                data = data.plus(warningHrHigh.getDataBytes())
            }
            9 -> {
                data = data.plus(warningHrLow.getDataBytes())
            }
        }
        return data
    }

    // 报警提示：呼吸暂停
    class WarningApnea() {
        var apnea = 0  // 0: Off 范围：10s/20s/30s；级别：高
        // reserved 3
        constructor(bytes: ByteArray) : this() {
            var index = 0
            apnea = byte2UInt(bytes[index])
        }
        fun getDataBytes(): ByteArray {
            return byteArrayOf(apnea.toByte())
                .plus(ByteArray(3))
        }
    }
    // 报警提示：漏气量高
    class WarningLeak() {
        var high = 0  // 0: Off 范围：15s/30s/45s/60s； 级别：中
        // reserved 3
        constructor(bytes: ByteArray) : this() {
            var index = 0
            high = byte2UInt(bytes[index])
        }
        fun getDataBytes(): ByteArray {
            return byteArrayOf(high.toByte())
                .plus(ByteArray(3))
        }
    }
    // 报警提示：潮气量低
    class WarningVt() {
        var low = 0  // 0: Off 范围：200-2000ml；步进：10ml；级别：中
        // reserved 2
        constructor(bytes: ByteArray) : this() {
            var index = 0
            low = toUInt(bytes.copyOfRange(index, index+2))
        }
        fun getDataBytes(): ByteArray {
            return int2ByteArray(low)
                .plus(ByteArray(2))
        }
    }
    // 报警提示：分钟通气量低
    class WarningVentilation() {
        var low = 0  // 0: off 范围：1-25L/min；步进：1L/min；级别：中
        // reserved 3
        constructor(bytes: ByteArray) : this() {
            var index = 0
            low = byte2UInt(bytes[index])
        }
        fun getDataBytes(): ByteArray {
            return byteArrayOf(low.toByte())
                .plus(ByteArray(3))
        }
    }
    // 报警提示：呼吸频率高
    class WarningRrHigh() {
        var high = 0  // 0:Off 范围：1-60bmp；步进：1bpm；级别：中
        // reserved 3
        constructor(bytes: ByteArray) : this() {
            var index = 0
            high = byte2UInt(bytes[index])
        }
        fun getDataBytes(): ByteArray {
            return byteArrayOf(high.toByte())
                .plus(ByteArray(3))
        }
    }
    // 报警提示：呼吸频率低
    class WarningRrLow() {
        var low = 0  // 0:Off 范围：1-60bmp；步进：1bpm；级别：中
        // reserved 3
        constructor(bytes: ByteArray) : this() {
            var index = 0
            low = byte2UInt(bytes[index])
        }
        fun getDataBytes(): ByteArray {
            return byteArrayOf(low.toByte())
                .plus(ByteArray(3))
        }
    }
    // 报警提示：血氧饱和度低
    class WarningSpo2Low() {
        var low = 0  // 0:Off 范围：80-95%；步进：1%；级别：中
        // reserved 3
        constructor(bytes: ByteArray) : this() {
            var index = 0
            low = byte2UInt(bytes[index])
        }
        fun getDataBytes(): ByteArray {
            return byteArrayOf(low.toByte())
                .plus(ByteArray(3))
        }
    }
    // 报警提示：脉率/心率高
    class WarningHrHigh() {
        var high = 0  // 0:Off 范围：100-240bmp；步进：10bpm；级别：中
        // reserved 3
        constructor(bytes: ByteArray) : this() {
            var index = 0
            high = byte2UInt(bytes[index])
        }
        fun getDataBytes(): ByteArray {
            return byteArrayOf(high.toByte())
                .plus(ByteArray(3))
        }
    }
    // 报警提示：脉率/心率低
    class WarningHrLow() {
        var low = 0  // 0:Off 范围：30-70bmp；步进：5bpm；级别：中
        // reserved 3
        constructor(bytes: ByteArray) : this() {
            var index = 0
            low = byte2UInt(bytes[index])
        }
        fun getDataBytes(): ByteArray {
            return byteArrayOf(low.toByte())
                .plus(ByteArray(3))
        }
    }
}